### PR TITLE
fix: migrate per-folder org settings from v2.31.0 to LS model on upgrade [IDE-2259]

### PR DESCRIPTION
**Pull request by nick-y-snyk**
## Summary
- `snyk.advanced.organization` moved from `resource`- to `window`-scoped and `autoSelectOrganization` was dropped, so a multi-root workspace's legacy per-folder `.vscode/settings.json` org values became inert on upgrade — folders silently fell back to auto-determined org.
- Reads the legacy per-folder values off disk once at activation, promotes them into `preferred_org`/`org_set_by_user`, before `LanguageServer` starts so the first `initializationOptions` already carries the migrated org.
- Gated one-shot by a `globalState` memento.

Jira: https://snyksec.atlassian.net/browse/IDE-2259

## Test plan
- [x] New unit tests in `folderOrgMigration.test.ts`: repro test against real `LanguageServerSettings.resolveFolderConfigs` shows org lost pre-fix, then confirms fix preserves it, plus edge cases (autoSelectOrganization opt-out, no legacy settings, merge with existing folder settings, one-shot memento gate, no premature LS notification).
- [x] `npm run test:unit` — full suite passes (pre-existing unrelated 2 failures on `cliExecutable`/`configuration` CLI-path tests, untouched by this diff).
- [x] `npm run lint:fix` — clean.
- [x] `npm run build` — clean.

### CODE REVIEW

File: src/snyk/common/configuration/folderOrgMigration.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    return {};
+  }
+}
+
+export async function migrateFolderOrgSettingsIfNeeded(
+  workspace: IVSCodeWorkspace,
+  configuration: IConfiguration,
+  context: ExtensionContext,
+): Promise<void> {
+  if (context.globalState.get(MEMENTO_FOLDER_ORG_MIGRATION_V1) === true) {
```
Comment (basti-snyk) [inline src/snyk/common/configuration/folderOrgMigration.ts]: **Critical — installation-global memento gates a per-workspace migration, so it runs for only ONE workspace ever.**

`context.globalState` is shared across every window/workspace of the extension install, but the data being recovered (legacy per-folder `.vscode/settings.json` org values) is per-folder and scattered across whatever projects the user opens over time. The migration only iterates the currently-open `getWorkspaceFolders()`, then burns this global flag (`:74`).

Consequences:
- User opens project A after upgrade → A migrates, memento set. Later opens project B in a separate window (its own legacy org) → this gate returns early → B's org is never recovered and is now inert (setting is window-scoped). Silent data loss — the exact "org lost on upgrade" bug this PR targets, still present for the common multi-project developer.
- Worse: if the first post-upgrade window has no folder / no legacy org, `:74` still sets the flag → *every* future project loses its org.

Root cause: wrong memento scope. Fix: gate on `context.workspaceState` (each workspace migrates once on first post-upgrade open), or store a `Set<string>` of migrated folder paths in globalState and migrate a folder the first time it's ever seen. The test that asserts "only runs once, gated by the globalState memento" encodes the wrong behavior; add a case proving a second, distinct workspace still migrates. — AI review

Comment (nick-y-snyk) [reply src/snyk/common/configuration/folderOrgMigration.ts]: Fixed in 6a69fb96 — the memento now tracks a `Set<string>` of already-checked folder paths (persisted as `string[]`) instead of a single install-wide boolean, so each distinct workspace folder migrates exactly once regardless of window/instance. Updated the existing 'only runs once' test and added a new test proving a never-before-seen folder still migrates when a different folder is already recorded.

File: src/snyk/common/configuration/folderOrgMigration.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+  for (const folder of foldersToCheck) {
+    const folderPath = folder.uri.fsPath;
+    // Record the folder as checked regardless of outcome, so a folder with no legacy org
+    // isn't re-read on every activation.
+    migratedFolderPaths.add(folderPath);
+
+    const legacy = await readLegacyOrgSettings(folderPath, logger);
+    // autoSelectOrganization:true is an explicit opt-out of the per-folder org — leave default.
+    if (!legacy.organization || legacy.autoSelectOrganization === true) {
```
Comment (andrewrobinsonhodges-snyk) [inline src/snyk/common/configuration/folderOrgMigration.ts]: [Should Fix] `migrateFolderOrgSettingsIfNeeded` skips migration whenever `autoSelectOrganization === true`, but in v2.31.0 an explicitly configured org always overrode auto-select (the setting's own docs: "If an organization is configured manually, this feature will be overridden"). The default also flipped from `false` (introduced in 47c6a2ac) to `true` (1edc9636) before removal, so early adopters could have both an explicit org and an explicit `autoSelectOrganization: true` in their folder settings — for them the explicit org was in effect, yet this clause silently drops it on upgrade, contradicting the "no silent org change" goal.

Fix: migrate whenever a legacy org is present, regardless of auto-select — the explicit org always won.
```ts
if (!legacy.organization) {
  continue;
}
```
(If skipping-on-auto-select is a deliberate product decision, worth calling out explicitly, since it diverges from v2.31.0 precedence.)

— AI review

File: src/snyk/common/configuration/folderOrgMigration.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+interface LegacyOrgSettings {
+  organization?: string;
+  autoSelectOrganization?: boolean;
+}
+
+async function readLegacyOrgSettings(folderPath: string, logger: ILog): Promise<LegacyOrgSettings> {
+  try {
+    const settingsPath = path.join(folderPath, '.vscode', 'settings.json');
+    const content = await fs.promises.readFile(settingsPath, 'utf-8');
+    const settings = parse(content) as Record<string, unknown> | undefined;
```
Comment (andrewrobinsonhodges-snyk) [inline src/snyk/common/configuration/folderOrgMigration.ts]: [Should Fix] A transiently malformed `.vscode/settings.json` causes permanent org loss. `parse()` is called without an errors collector, so a hard syntax error returns `undefined`; the `!settings` guard returns `{}` (no org), but line 69 still records the folder as migrated, so it is never re-read. A user whose settings file was momentarily broken during upgrade loses their per-folder org for good.

Fix: pass an errors array and, on a genuine parse failure (distinct from ENOENT / empty file), signal the caller to NOT record the folder as migrated so it retries next activation.
```ts
const errors: ParseError[] = [];
const settings = parse(content, errors) as Record<string, unknown> | undefined;
if (errors.length > 0) {
  // real syntax error — surface so the folder is retried, not marked migrated
}
```

— AI review

File: src/snyk/common/configuration/folderOrgMigration.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  configuration: IConfiguration,
+  context: ExtensionContext,
+  logger: ILog,
+): Promise<void> {
+  // Tracked per-folder-path (not a single install-wide boolean): globalState is shared across
+  // every window of this extension install, but the legacy org lives per workspace folder. A
+  // per-install flag would let the first window post-upgrade permanently skip every other
+  // folder/window that hasn't been seen yet (IDE-2259 review fix).
+  const migratedFolderPaths = new Set(context.globalState.get<string[]>(MEMENTO_FOLDER_ORG_MIGRATION_V1) ?? []);
+  const foldersToCheck = workspace.getWorkspaceFolders().filter(folder => !migratedFolderPaths.has(folder.uri.fsPath));
```
Comment (andrewrobinsonhodges-snyk) [inline src/snyk/common/configuration/folderOrgMigration.ts:112]: [Should Fix] Only each folder's `.vscode/settings.json` is read, so a `resource`-scoped org set in a multi-root `.code-workspace` file (top-level `settings` or `folders[].settings`) is not migrated — those users silently revert to auto-determined org.

Fix: decide explicitly whether workspace-file orgs are in scope. If out of scope, add a code comment documenting the gap; if in scope, also read the `.code-workspace` settings for each folder.

— AI review

File: src/snyk/common/configuration/folderOrgMigration.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    const settingsPath = path.join(folderPath, '.vscode', 'settings.json');
+    const content = await fs.promises.readFile(settingsPath, 'utf-8');
+    const settings = parse(content) as Record<string, unknown> | undefined;
+    if (!settings) {
+      return {};
+    }
+    const organization = settings[ADVANCED_ORGANIZATION];
+    return {
+      organization: typeof organization === 'string' ? organization : undefined,
+      autoSelectOrganization: settings[ADVANCED_AUTO_SELECT_ORGANIZATION] as boolean | undefined,
```
Comment (andrewrobinsonhodges-snyk) [inline src/snyk/common/configuration/folderOrgMigration.ts]: [Suggestion] `autoSelectOrganization` is cast `as boolean | undefined` with no runtime guard, unlike the adjacent `organization` which got a `typeof === "string"` guard. It is safe (a non-boolean can't match `=== true`), but the cast lies to the type system and the guard commit only half-applies.

Fix: mirror the org guard for symmetry.
```ts
const autoSelect = settings[ADVANCED_AUTO_SELECT_ORGANIZATION];
return {
  organization: typeof organization === "string" ? organization : undefined,
  autoSelectOrganization: typeof autoSelect === "boolean" ? autoSelect : undefined,
};
```

— AI review

File: src/snyk/common/configuration/folderOrgMigration.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  if (migrated) {
+    // triggerConfigChangeEvent=false: this runs before LanguageServer/LanguageClient exist,
+    // so there's nothing to notify — the migrated configs are picked up directly by the
+    // initial resolveFolderConfigs() call when building initializationOptions.
+    await configuration.setFolderConfigs(Array.from(configsByPath.values()), false);
+  }
+
+  await context.globalState.update(MEMENTO_FOLDER_ORG_MIGRATION_V1, Array.from(migratedFolderPaths));
```
Comment (andrewrobinsonhodges-snyk) [inline src/snyk/common/configuration/folderOrgMigration.ts:158]: [Suggestion] `MEMENTO_FOLDER_ORG_MIGRATION_V1` accumulates every folder path ever opened and is never pruned when a folder leaves the workspace. Harmless at realistic scale (only Set membership is read), but it grows monotonically in globalState. Worth noting; no action needed unless you want to prune paths no longer in the workspace.

— AI review

File: src/snyk/common/configuration/folderOrgMigration.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  workspace: IVSCodeWorkspace,
+  configuration: IConfiguration,
+  context: ExtensionContext,
+  logger: ILog,
+): Promise<void> {
+  // Tracked per-folder-path (not a single install-wide boolean): globalState is shared across
+  // every window of this extension install, but the legacy org lives per workspace folder. A
+  // per-install flag would let the first window post-upgrade permanently skip every other
+  // folder/window that hasn't been seen yet (IDE-2259 review fix).
+  const migratedFolderPaths = new Set(context.globalState.get<string[]>(MEMENTO_FOLDER_ORG_MIGRATION_V1) ?? []);
```
Comment (andrewrobinsonhodges-snyk) [inline src/snyk/common/configuration/folderOrgMigration.ts:111]: [Suggestion] `globalState` is shared across all windows of this install, so two windows activating concurrently can read-modify-write the migrated-paths list with last-write-wins, potentially dropping a path. Worst case is one extra re-read of a folder's `settings.json` on the next activation — negligible, noting for completeness.

— AI review

File: src/snyk/common/configuration/folderOrgMigration.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+ * only this on-disk read can still recover them (IDE-2259).
+ */
+interface LegacyOrgSettings {
+  organization?: string;
+  autoSelectOrganization?: boolean;
+}
+
+async function readLegacyOrgSettings(folderPath: string, logger: ILog): Promise<LegacyOrgSettings> {
+  try {
+    const settingsPath = path.join(folderPath, '.vscode', 'settings.json');
```
Comment (andrewrobinsonhodges-snyk) [inline src/snyk/common/configuration/folderOrgMigration.ts]: [Suggestion] Reading `.vscode/settings.json` directly is a justified boundary crossing — the settings are deregistered so the VS Code config API can no longer return them — but it is the only place the extension depends on VS Code's on-disk settings layout.

Fix: a short comment pointing at that coupling would help if VS Code ever changes the location.

— AI review

File: src/snyk/common/configuration/folderOrgMigration.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+export async function migrateFolderOrgSettingsIfNeeded(
+  workspace: IVSCodeWorkspace,
+  configuration: IConfiguration,
+  context: ExtensionContext,
+  logger: ILog,
+): Promise<void> {
+  // Tracked per-folder-path (not a single install-wide boolean): globalState is shared across
+  // every window of this extension install, but the legacy org lives per workspace folder. A
+  // per-install flag would let the first window post-upgrade permanently skip every other
+  // folder/window that hasn't been seen yet (IDE-2259 review fix).
```
Comment (andrewrobinsonhodges-snyk) [inline src/snyk/common/configuration/folderOrgMigration.ts:110]: Not sure we need the reference to IDE-2259 here

File: src/snyk/common/constants/settings.ts [hunk 10]
```diff
@@ -15,6 +15,9 @@ export const ADVANCED_ADDITIONAL_PARAMETERS_SETTING = `${CONFIGURATION_IDENTIFIE
 export const ADVANCED_ADDITIONAL_ENVIRONMENT_SETTING = `${CONFIGURATION_IDENTIFIER}.advanced.additionalEnvironment`;
 export const ADVANCED_CUSTOM_ENDPOINT = `${CONFIGURATION_IDENTIFIER}.advanced.customEndpoint`;
 export const ADVANCED_ORGANIZATION = `${CONFIGURATION_IDENTIFIER}.advanced.organization`;
+// Removed from package.json (scope resource -> window); kept as a string constant
```
Comment (andrewrobinsonhodges-snyk) [inline src/snyk/common/constants/settings.ts:18]: I don't like the AI's habit of explaining the history in comments

**Review by andrewrobinsonhodges-snyk — APPROVED** [review]

File: src/snyk/common/configuration/folderOrgMigration.ts (hunk 5 above)
Comment (nick-y-snyk) [reply src/snyk/common/configuration/folderOrgMigration.ts]: Skipping — this cast is provably safe (a non-boolean value can never satisfy `=== true`), and it doesn't share the risk profile of the string cast (which could forward a non-string into org-selection logic). Not worth the diff.

File: src/snyk/common/configuration/folderOrgMigration.ts (hunk 2 above)
Comment (nick-y-snyk) [reply src/snyk/common/configuration/folderOrgMigration.ts]: Deliberate — won't change. Even though v2.31.0 gave the explicit org precedence over auto-select, we want the migration to respect `autoSelectOrganization: true` as the user's intent going forward and skip seeding a stale explicit org in that case. Not a bug, a product-behavior call for this migration.

File: src/snyk/common/configuration/folderOrgMigration.ts (hunk 8 above)
Comment (nick-y-snyk) [reply src/snyk/common/configuration/folderOrgMigration.ts]: Added a comment on `readLegacyOrgSettings` noting the deliberate coupling to VS Code's on-disk settings.json layout, since the config API can't return a deregistered resource-scoped setting. Commit `64271e71`.

File: src/snyk/common/configuration/folderOrgMigration.ts (hunk 3 above)
Comment (nick-y-snyk) [reply src/snyk/common/configuration/folderOrgMigration.ts]: Fixed — `readLegacyOrgSettings` now passes a `ParseError[]` collector to `parse()`. On a genuine syntax error, the caller skips adding the folder to `migratedFolderPaths`, so it's retried next activation; ENOENT and valid-empty-file cases are unchanged. Added tests for all three outcomes. Commit `17f00a18`.

File: src/snyk/common/configuration/folderOrgMigration.ts (hunk 4 above)
Comment (nick-y-snyk) [reply src/snyk/common/configuration/folderOrgMigration.ts:112]: Fixed — added `getWorkspaceFile()` to `IVSCodeWorkspace`. The migration now reads a `.code-workspace` file's top-level `settings` once and uses its org as a fallback only for folders with no org of their own (folder-level settings still win, matching real VS Code precedence). Remote/virtual workspace files are left as a noted limitation for now — `IVSCodeWorkspace` doesn't expose `vscode.workspace.fs` yet. Added tests for the fallback and the folder-wins case. Commit `03bc8147`.
